admin/tools/provision/cloudmanager.py

```diff
@@ -0,0 +1,311 @@
+        self.snapshot_name = config.get('openstack', 'snapshot_name')
+        self.image = self.nova.images.find(name=base_image_name)
+
+        flavor_name = config.get('openstack', 'flavor_name')
+        self.flavor = self.nova.flavors.find(name=flavor_name)
+
+        self.network_name = config.get('openstack', 'network_name')
+        if config.get('openstack', 'net-id'):
+            unicode_net_id = unicode(config.get('openstack', 'net-id'))
+            self.nics = [{'net-id': unicode_net_id}]
+        else:
+            self.nics = []
+        self.ssh_security_group = config.get('openstack', 'ssh_security_group')
+
+        # Upload ssh public key
+        if add_ssh_key:
+            self.key = "{}-qserv".format(self._safe_username)
+        else:
+            self.key = None
+
+        self.key_filename = '~/.ssh/id_rsa'
+
+
+
+    def get_nb_servers(self):
+        """
+        Get number of servers to boot
+        """
+        args = None
+        return args.nbServers
+
+    def nova_image_create(self, instance):
+        """
+        Create an openstack image containing Docker
+        """
+        logging.info("Creating Qserv image")
+        qserv_image = instance.create_image(self.snapshot_name)
+        status = None
+        while status != 'ACTIVE':
+            time.sleep(5)
+            status = self.nova.images.get(qserv_image).status
+        logging.info("SUCCESS: Qserv image '{}' is active".format(self.snapshot_name))
+
+    def nova_servers_create(self, instance_id, userdata):
+        """
+        Boot an instance and check status
+        """
+        instance_name = "{0}-qserv-{1}".format(self._safe_username, instance_id)
+        logging.info("Launch an instance {}".format(instance_name))
+
+        # Launch an instance from an image
+        instance = self.nova.servers.create(name=instance_name,
+                                            image=self.image,
+                                            flavor=self.flavor,
+                                            userdata=userdata,
+                                            key_name=self.key,
+                                            nics=self.nics)
+        # Poll at 5 second intervals, until the status is no longer 'BUILD'
+        status = instance.status
+        while status == 'BUILD':
+            time.sleep(5)
+            instance.get()
+            status = instance.status
+        logging.info("status: {}".format(status))
+        logging.info("Instance {} is active".format(instance_name))
+
+        return instance
+
+    def detect_end_cloud_config(self, instance):
+        """
+        Add clean wait for cloud-init completion
+        """
+        check_word = "---SYSTEM READY FOR SNAPSHOT---"
+        end_word = None
+        while not end_word:
+            time.sleep(10)
+            output = instance.get_console_output()
+            logging.debug("console output: {}".format(output))
```

Comment:
For `logging` please use lazy formatting:

    logging.debug("console output: %s", output)

which avoids unnecessary formatting in case the output is discarded. This applies to all logging messages in this module.